discovery-provider/src/queries/get_collection_library.py

```diff
@@ -0,0 +1,201 @@
+    add_query_pagination,
+    populate_playlist_metadata,
+)
+from src.utils import helpers
+from src.utils.db_session import get_db_read_replica
+
+
+class CollectionType(str, enum.Enum):
+    playlist = "playlist"
+    album = "album"
+
+
+class GetCollectionLibraryArgs(TypedDict):
+    # The current user logged in (from route param)
+    user_id: int
+
+    collection_type: CollectionType
+
+    # The maximum number of listens to return
+    limit: int
+
+    # The offset for the listen history
+    offset: int
+
+    filter_type: LibraryFilterType
+
+    # Optional filter for the returned results
+    query: Optional[str]
+
+    sort_direction: Optional[SortDirection]
+    sort_method: Optional[LimitedSortMethod]
+
+
+# Most of the code in this file is copied from/similar to src/queries/get_track_library.py -
+# see that file for more in depth comments about query structure etc
+def _get_collection_library(args: GetCollectionLibraryArgs, session):
+    user_id = args["user_id"]
+    limit = args["limit"]
+    offset = args["offset"]
+    collection_type = args["collection_type"]
+    filter_type = args["filter_type"]
+
+    query = args.get("query")
+    sort_method = args.get("sort_method", LimitedSortMethod.added_date)
+    sort_direction = args.get("sort_direction", SortDirection.desc)
+
+    # Doesn't yet support album/playlist purchases
+    if not filter_type or filter_type == LibraryFilterType.purchase:
+        raise ValueError("Invalid filter type")
+
+    sort_fn = desc if sort_direction == SortDirection.desc else asc
+
+    own_playlists_base = session.query(
+        Playlist, Playlist.created_at.label("item_created_at")
+    ).filter(
+        Playlist.is_current == True,
+        Playlist.is_album == (collection_type == CollectionType.album),
+        Playlist.playlist_owner_id == user_id,
+    )
+
+    favorites_base = (
+        session.query(Playlist, Save.created_at.label("item_created_at"))
+        .join(Save, Save.save_item_id == Playlist.playlist_id)
+        .filter(
+            Playlist.is_current == True,
+            Playlist.is_private == False,
+            Playlist.is_album == (collection_type == CollectionType.album),
+            Save.user_id == user_id,
+            Save.is_current == True,
+            Save.is_delete == False,
+            Save.save_type == SaveType[collection_type],
+        )
+    )
+
+    reposts_base = (
+        session.query(Playlist, Repost.created_at.label("item_created_at"))
+        .join(Repost, Repost.repost_item_id == Playlist.playlist_id)
+        .filter(
+            Playlist.is_current == True,
+            Playlist.is_private == False,
+            Playlist.is_album == (collection_type == CollectionType.album),
+            Repost.user_id == user_id,
+            Repost.is_current == True,
+            Repost.is_delete == False,
+            Repost.repost_type == RepostType[collection_type],
+        )
+    )
+
+    # Favorites are unioned with own playlists by design
```

Comment:
Interesting... wonder if we should include own tracks in the tracks library?

Reply:
Would be a pretty drastic UI divergence from what we have today... where for playlists, we already show favorited playlists mixed with own playlists

Reply:
Still odd though, how do you download your own tracks for offline listening 🤔

Reply:
Maybe eventually artist dashboard has a download toggle, or migrates wholesale to library...